Bluetooth: Host: Enable ENTROPY_GENERATOR for BT_HOST_CRYPTO_PRNG

jhedberg opened the pull request:
With the recent change to PSA crypto, it's important to enable the top-level entropy driver Kconfig option, so that the underlying entropy driver can be enabled when necessary. The one exception to this is if a test (i.e. insecure) random number generator is enabled, since those will also provide the underlying z_impl_sys_rand_get() implementation, which is what's ultimately required by PSA.

Fixes #82344

jhedberg committed:
Bluetooth: Host: Enable ENTROPY_GENERATOR for BT_HOST_CRYPTO_PRNG

With the recent change to PSA crypto, it's important to enable the
top-level entropy driver Kconfig option, so that the underlying entropy
driver can be enabled when necessary. The one exception to this is if a
test (i.e. insecure) random number generator is enabled, since those will
also provide the underlying z_impl_sys_rand_get() implementation, which is
what's ultimately required by PSA.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg commented:
@carlescufi @ppryga-nordic @wilu-nordic is it intentional that nRF54 is missing a `zephyr,entropy` DTS chosen property? All other Nordic boards seem to have it. That seems to be the reason why CI in this PR is failing.

jhedberg commented:
> @carlescufi @ppryga-nordic @wilu-nordic is it intentional that nRF54 is missing a `zephyr,entropy` DTS chosen property? All other Nordic boards seem to have it. That seems to be the reason why CI in this PR is failing.

I did a little bit research into this, however, I'm not sure how to fix it. There are several open questions:
1. The board that's failing is `nrf54l15dk/nrf54l15/cpuapp`, however despite being an app core, it's actually enabling the native controller instead of e.g. the IPC HCI driver. Why is that? https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/dts/arm/nordic/nrf54l_05_10_15_cpuapp.dtsi#L17-L19
2. The DTS for this board does have a `rng_hci` node, but that doesn't help since the native controller has a hardcoded expectation of something called `rng` instead of relying on the chosen node. There's even a FIXME in the controller for this: https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L61-L66

So, could someone from Nordic help out a bit here? What's the right fix?

jhedberg commented:
To the previous comment, I also wanted to add that the native controller builds fine with current main by pure luck. Since no entropy driver is enabled (dues to `CONFIG_ENTROPY_GENERATOR=y`) it means that `CONFIG_ENTROPY_DRIVER` is never selected, and so the above snippet from the controller `lll.c` will never actually be included in the build.

ppryga-nordic commented:
> > @carlescufi @ppryga-nordic @wilu-nordic is it intentional that nRF54 is missing a `zephyr,entropy` DTS chosen property? All other Nordic boards seem to have it. That seems to be the reason why CI in this PR is failing.
> 
> I did a little bit research into this, however, I'm not sure how to fix it. There are several open questions:
> 
>     1. The board that's failing is `nrf54l15dk/nrf54l15/cpuapp`, however despite being an app core, it's actually enabling the native controller instead of e.g. the IPC HCI driver. Why is that? https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/dts/arm/nordic/nrf54l_05_10_15_cpuapp.dtsi#L17-L19
> 
>     2. The DTS for this board has both `rng_hci` and `psa_rng` nodes, but that doesn't help since the native controller has a hardcoded expectation of something called `rng` instead of relying on the chosen node. There's even a FIXME in the controller for this: https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L61-L66
> 
> 
> So, could someone from Nordic help out a bit here? What's the right fix?

1. The nRF54L15 has no `network core` hence full stack is build for the app core. 
2. @carlescufi @joerchan it looks like the changed done here in downstream: https://github.com/nrfconnect/sdk-zephyr/commit/de051d6ab581c468f499fa385c942a5067962ef4#diff-25882a5f282f0438ba2a052356fa5ccf9a6f9ba00d45affe76aa048c7bf7bb54R19 is missing or at least its part. Also it looks like @jhedberg  is right about the [zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c](https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L61-L66). Shouldn't it use `DT_CHOSEN` @cvinayak ?

joerchan commented:
> Shouldn't it use DT_CHOSEN

That was done on purpose, so nrf53 / nrf52 devices would always choose the RNG even with cryptocell present because that driver requires a lot of stack-space and is called from all of the bt threads.
https://github.com/zephyrproject-rtos/zephyr/commit/f756b791248b40a65b0e5d0426b417cd3180aa9f

jhedberg commented:
@joerchan what would you propose as a fix here then? Right now enabling CONFIG_ENTROPY_GENERATOR will cause a build failure for this board, which seems like a clear bug to me.

Vge0rge commented:
> > > @carlescufi @ppryga-nordic @wilu-nordic is it intentional that nRF54 is missing a `zephyr,entropy` DTS chosen property? All other Nordic boards seem to have it. That seems to be the reason why CI in this PR is failing.
> > 
> > 
> > I did a little bit research into this, however, I'm not sure how to fix it. There are several open questions:
> > ```
> > 1. The board that's failing is `nrf54l15dk/nrf54l15/cpuapp`, however despite being an app core, it's actually enabling the native controller instead of e.g. the IPC HCI driver. Why is that? https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/dts/arm/nordic/nrf54l_05_10_15_cpuapp.dtsi#L17-L19
> > 
> > 2. The DTS for this board has both `rng_hci` and `psa_rng` nodes, but that doesn't help since the native controller has a hardcoded expectation of something called `rng` instead of relying on the chosen node. There's even a FIXME in the controller for this: https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L61-L66
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, could someone from Nordic help out a bit here? What's the right fix?
> 
> 1. The nRF54L15 has no `network core` hence full stack is build for the app core.
> 2. @carlescufi @joerchan it looks like the changed done here in downstream: [nrfconnect/sdk-zephyr@de051d6#diff-25882a5f282f0438ba2a052356fa5ccf9a6f9ba00d45affe76aa048c7bf7bb54R19](https://github.com/nrfconnect/sdk-zephyr/commit/de051d6ab581c468f499fa385c942a5067962ef4#diff-25882a5f282f0438ba2a052356fa5ccf9a6f9ba00d45affe76aa048c7bf7bb54R19) is missing or at least its part. Also it looks like @jhedberg  is right about the [zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c](https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L61-L66). Shouldn't it use `DT_CHOSEN` @cvinayak ?

2. The psa_rng is enabled downstream because the hardware which supports RNG and crypto in nRF54l15 (CRACEN) is currently only supported downstream. So currently in upstream zephyr enabling entropy generator should fail to build for the nRF54l15.  Until we add support for this it makes sense to me to disable building the samples/apps which require entropy generator enabled.

jhedberg commented:
> So currently in upstream zephyr enabling entropy generator should fail to build for the nRF54l15.  Until we add support for this it makes sense to me to disable building the samples/apps which require entropy generator enabled.

That implies that the premise of this pull request is wrong, i.e. it's not actually `BT_HOST_CRYPT_PRNG` that requires `ENTROPY_GENERATOR`, rather something else. I.e. what's making the build succeed for boards like nRF54, but making it fail for others, as shown e.g. in #82344.

jhedberg commented:
Answering myself. What's ultimately required by PSA/mbedtls is an implementation of ` z_impl_sys_rand_get()`. To get a secure version of that requires in practice `CONFIG_ENTROPY_GENERATOR=y`. However, current nRF54 upstream enables `CONFIG_TEST_RANDOM_GENERATOR=y` which translates to `CONFIG_TIMER_RANDOM_GENERATOR=y` and that driver also implements `z_impl_sys_rand_get()`, which is why CI ends up passing.

jhedberg commented:
I changed the select to:
```
select ENTROPY_GENERATOR if !TEST_RANDOM_GENERATOR
```

It's not super intuitive (and I'm not sure in the long run the right fix), but it should make CI happy now for nRF54 as well.

jhedberg commented:
> I changed the select to:
> 
> ```
> select ENTROPY_GENERATOR if !TEST_RANDOM_GENERATOR
> ```
> 
> It's not super intuitive (and I'm not sure in the long run the right fix), but it should make CI happy now for nRF54 as well.

Actually, no, that doesn't seem to work either - seems like it's causing some circular dependency or "impossible dependency chain" error with Kconfig.

carlescufi commented:
@jhedberg @joerchan @ppryga-nordic we are currently working on a plan to try and get RNG support for the nRF54L series in upstream Zephyr. If that succeeds then we should be able to treat the nRF54L as the nRF53 or nRF52. Copying @aescolar as well here.

joerchan commented:
> Actually, no, that doesn't seem to work either - seems like it's causing some circular dependency or "impossible dependency chain" error with Kconfig.

@jhedberg I think the select is wrong here, when using PSA you'd want to use the PSA API psa_want_generate_random.
If that function requires entropy or something else is a different story. But you should not be demanding entropy from this configuration.

Also `z_impl_sys_csrand_get` is provided by the random subsystem, not the entropy drivers, so select of entropy is in any case wrong also.

jhedberg commented:
> > Actually, no, that doesn't seem to work either - seems like it's causing some circular dependency or "impossible dependency chain" error with Kconfig.
> 
> @jhedberg I think the select is wrong here, when using PSA you'd want to use the PSA API psa_want_generate_random. If that function requires entropy or something else is a different story. But you should not be demanding entropy from this configuration.
> 
> Also `z_impl_sys_csrand_get` is provided by the random subsystem, not the entropy drivers, so select of entropy is in any case wrong also.

Thanks @joerchan. The initial reaction to this issue (that surfaced with the move to PSA) was that HCI drivers started trying to add `select ENTROPY_GENERATOR` to their Kconfig options, which I think is even more wrong than trying to do this where the dependency on crypto/randomness actually exists.

While it's clear from your comment what shouldn't be done, it's still not clear to me what the right fix is (in concrete terms). Could you help clarify this? Thanks.

jhedberg commented:
> > > Actually, no, that doesn't seem to work either - seems like it's causing some circular dependency or "impossible dependency chain" error with Kconfig.
> > 
> > 
> > @jhedberg I think the select is wrong here, when using PSA you'd want to use the PSA API psa_want_generate_random. If that function requires entropy or something else is a different story. But you should not be demanding entropy from this configuration.
> > Also `z_impl_sys_csrand_get` is provided by the random subsystem, not the entropy drivers, so select of entropy is in any case wrong also.
> 
> Thanks @joerchan. The initial reaction to this issue (that surfaced with the move to PSA) was that HCI drivers started trying to add `select ENTROPY_GENERATOR` to their Kconfig options, which I think is even more wrong than trying to do this where the dependency on crypto/randomness actually exists (i.e. what this PR attempted to do).
> 
> While it's clear from your comment what shouldn't be done, it's still not clear to me what the right fix is (in concrete terms). Could you help clarify this? Thanks.

frkv commented:
@valeriosetti made PSA crypto support RNG from sources that are not **C**ryptographically **S**ecure. I think this would be relevant for the ongoing case

https://github.com/zephyrproject-rtos/zephyr/pull/76408

valeriosetti commented:
FWIW I faced this issue with NRF54 devices when working on the TinyCrypt deprecation PR, and I left a comment here:
https://github.com/valeriosetti/zephyr/blob/98e6ea865ffe0df45d7869f10ba093c2b5e45dba/samples/bluetooth/beacon/boards/nrf54l15dk_nrf54l15_cpuapp.conf#L1-L2

valeriosetti commented:
@jhedberg as a "proof" of the change introduced by this PR, do you think it's worth reverting all the additions I did in this commit: https://github.com/zephyrproject-rtos/zephyr/commit/55c0b2ea45970bfb81dc3b61abe8fc94c40bd6b3#diff-9d910f456258e71150cba29ad4c0b2dad8f742fe66a8c73353118918e380c093?

valeriosetti commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -179,6 +179,8 @@ config BT_HOST_CRYPTO
 	default y
+	# An implementation of z_impl_sys_rand_get() is required
+	select ENTROPY_GENERATOR if !TEST_RANDOM_GENERATOR
 	select PSA_WANT_ALG_SHA_256
```
I know this is a comment on a pre-existing code (that I did btw), but since you are at this I think we can drop `PSA_WANT` below because HMAC-DRBG was only used in TinyCrypt to generate random data, while with PSA Crypto API we call `psa_generate_random()` directly.
If you think that this does not belong to this PR, then I can open a separate one ;)

jhedberg replied:
Which one of the `PSA_WANT`? All of them?

valeriosetti replied:
Yes, all of them. The rationale is:
- if `psa_generate_random()` can rely on some entropy generator, then no crypto algorithm is required on top of that;
- if a valid entropy source is not available, `psa_generate_random()` will fall back to legacy mode (`MBEDTLS_PSA_CRYPTO_LEGACY_RNG`) where all the required legacy should be automatically selected.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -179,6 +179,8 @@ config BT_HOST_CRYPTO
 config BT_HOST_CRYPTO_PRNG
 	bool "Use PSA crypto API library for random number generation"
 	default y
+	# An implementation of z_impl_sys_rand_get() is required
```
The error I get when compiling is this:
```
/Users/johedber/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: modules/mbedtls/libmbedTLSBase.a(zephyr_entropy.c.obj): in function `sys_rand_get':
/Users/johedber/src/zephyr/zephyr/build/zephyr/include/generated/zephyr/syscalls/random.h:37: undefined reference to `z_impl_sys_rand_get'
```

So I think the comment is correct.

That said, I realized this is not the right place for the select. I get the error even when `BT_HOST_CRYPTO_PRNG=n` as long as `BT_HOST_CRYPTO=y`, so I've moved the select to the latter option.

valeriosetti replied:
Mmmm that's a bit strange. Did you enabled `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG`? Or perhaps you have `CONFIG_MBEDTLS_PSA_CRYPTO_LEGACY_RNG` enabled in your build?

jhedberg replied:
I'm not explicitly enabling anything. All I'm doing is trying out building Bluetooth applications for various boards where the HCI driver (or any other board-specific option) doesn't do an explicit `select ENTROPY_GENERATOR`. In the most recent case my build command was this (requires `west blobs fetch hal_infineon`):
```
west build -p -b cy8cproto_063_ble/cyble_416045_02 samples/bluetooth/observer -- -DCONFIG_BT_HOST_CRYPTO_PRNG=n
```
After checking `build/zephyr/.config` I can see that `CONFIG_MBEDTLS_PSA_CRYPTO_LEGACY_RNG=y`.

jhedberg committed:
Bluetooth: Host Remove unnecessary PSA_WANT_* selections

- If psa_generate_random() can rely on some entropy generator, then no
  crypto algorithm is required on top of that

- If a valid entropy source is not available, psa_generate_random() will
  fall back to legacy mode (MBEDTLS_PSA_CRYPTO_LEGACY_RNG) where all the
  required legacy should be automatically selected.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

erwango commented:
Following is still failing (`undefined reference to 'z_impl_sys_rand_get'`):
```
west twister -p nucleo_g0b1re/stm32g0b1xx -s samples/bluetooth/peripheral_hr/sample.bluetooth.peripheral_hr.frdm_kw41z_shield
```

jhedberg commented:
> west twister -p nucleo_g0b1re/stm32g0b1xx -s samples/bluetooth/peripheral_hr/sample.bluetooth.peripheral_hr.frdm_kw41z_shield



> Following is still failing (`undefined reference to 'z_impl_sys_rand_get'`):
> 
> ```
> west twister -p nucleo_g0b1re/stm32g0b1xx -s samples/bluetooth/peripheral_hr/sample.bluetooth.peripheral_hr.frdm_kw41z_shield
> ```

That implies that this platform doesn't have any entropy driver available, even with `ENTROPY_GENERATOR` enabled, similar to the situation with nRF54.

dleach02 commented:
I'll add that my experience with the RNG/Entropy Kconfig setup is that it is a bit odd. A number of years ago I added the cryptographically secure RNG and struggled with the Kconfigs around this (it was early in my Zephyr exposure). This feels like it should be set when pulling in PSA crypto. Although the next question would be whether we need the cryptographically secure RNG or not.

dleach02 commented:
ENTROPY_GENERATOR is the key to getting the RNG function and in some places we enable it based on other uses. For example, the nordic boards appear to turn it on when BT is true. But now that we have mbedtls module pulling in zephyr_entropy.c we need to turn on this config IF MBEDTLS_ENTROPY_POLL_ZEPHYR || MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG is true. But we create these dependency loops because we have places where ENTROPY_GENERATOR is set to true but depends on MBEDTLS.

Maybe we need to heavily rework the kconfig around ENTROPY and have this ENTROPY_GENERATOR only be set IF mbedtls is being used... or any additional item that wants to do RNG

dleach02 commented:
The challenge is that we set in the board Kconfig.defconfig the type of RNG to use but usually it seems we set it based on the setting of some other subsystem like "if NETWORKING". 

If we instead have the platforms specify which RNG to use IF RNG is needed then have the subsystem set the need for RNG that might be a better way

valeriosetti commented:
@dleach02 in #82497 I'm trying to enable `ENTROPY_GENERATOR` based on `zephyr,entropy` enabled in the device-tree. This was a suggestion made by @jhedberg on DIscord and I think it makes sense. There are some failures that I need to fix, but not that much after all

dleach02 commented:
While ENTROPY_GENERATOR is an issue here, the primary problem is that ENTROPY_GENERATOR is weirdly tied up with using the RNG function. There is a legit use of TEST_RANDOM that doesn't need ENTROPY_GENERATOR. 

ping @d3zd3z @ceolin

dleach02 commented:
Note that the modules/mbedtls/zephyr_entropy.c is going to want sys_rand_get() #if defined(CONFIG_MBEDTLS_ENTROPY_POLL_ZEPHYR) || defined(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG). The sys_rand_get() could be provided by random_timer.c or random_test_csprng.c where neither of these need ENTROPY_GENERATOR

jhedberg commented:
I'll close this as it's being superseded by #82497. There's likely going to be additional work to be done for platforms that don't have a proper entropy driver, but that's something to deal with on a board level and not Bluetooth, or if there is something needed on the Bluetooth side it's going to look very different from this PR.
